### PR TITLE
AWS: Provide option to use AWS KMS encryption

### PR DESCRIPTION
**Pull request by shermanyin**
Secrets used in the deployment can be optionally encrypted before being
entered into the .tfvars file.

Signed-off-by: Sherman Yin <syin@teradici.com>

### CODE REVIEW

File: docs/README-aws.md [hunk 1]
```diff
@@ -22,12 +22,26 @@ aws_secret_access_key = <your_key>
...
+### (Optional) Encrypting Secrets
+Secrets required as input to the Terraform scripts include Active Directory passwords, PCoIP registration key and the connector token. These secrets are stored in the local files terraform.tfvars and terraform.tfstate, and will also be uploaded as part of provisioning scripts to an AWS S3 bucket. Secrets may also show up in Terraform logs.
+
+The Terraform scripts are designed to support both plaintext and KMS-encrypted secrets. Plaintext secrets requires no extra steps, but will be stored in plaintext in the above mentioned locations. It is recommended that secrets are first encrypted before being entered into terraform.tfvars.
+
+To encrypt secrets using the KMS CMK created above, follow the instructions here: https://docs.aws.amazon.com/cli/latest/reference/kms/encrypt.html. Base64 encode the ciphertext before copying and pasting it into terraform.tfvars. For example, execute the following command in a Linux shell to get the base64-encoded ciphertext:
+
+```aws kms encrypt --key-id <cmk-id> --plaintext <secret> --output text --query CiphertextBlob```
+
+The following commnad can be used to decrypt the ciphertext:
```
Comment (Edwin-Pau) [inline docs/README-aws.md]: Typo on the word command

File: modules/aws/cac/cac-startup.sh.tmpl [hunk 2]
```diff
@@ -15,11 +15,22 @@ log() {
...
+        CAC_TOKEN=${cac_token}
+
+    else
+        log "Using encryption key ${customer_master_key_id}"
+
+        apt install -y awscli
+
+        PCOIP_REGISTRATION_CODE=$(aws kms decrypt --region ${aws_region} --ciphertext-blob fileb://<(echo "${pcoip_registration_code}" | base64 -d) --output text --query Plaintext | base64 -d)
+        SERVICE_ACCOUNT_PASSWORD=$(aws kms decrypt --region ${aws_region} --ciphertext-blob fileb://<(echo "${service_account_password}" | base64 -d) --output text --query Plaintext | base64 -d)
+        CAC_TOKEN=$(aws kms decrypt --region ${aws_region} --ciphertext-blob fileb://<(echo "${cac_token}" | base64 -d) --output text --query Plaintext | base64 -d)
```
Comment (Edwin-Pau) [inline modules/aws/cac/cac-startup.sh.tmpl:32]: This decryption is via curl POST using access token for GCP, should we convert that one to use gcloud?

File: modules/aws/dc/dc-startup.ps1.tmpl [hunk 3]
```diff
@@ -8,9 +8,28 @@ $LOG_FILE = "C:\Teradici\provisioning.log"
...
+    }
+}
+
 Start-Transcript -Path $LOG_FILE -Append -IncludeInvocationHeader
 
-"Not using encryption"
+if ([string]::IsNullOrWhiteSpace("${customer_master_key_id}")) {
+    "Not using encryption"
+} else {
+    "Using ecnryption key ${customer_master_key_id}"
```
Comment (Edwin-Pau) [inline modules/aws/dc/dc-startup.ps1.tmpl]: Typo on encryption

File: modules/aws/dc/new_domain_admin_user.ps1.tpl [hunk 4]
```diff
@@ -13,9 +13,28 @@ $LOG_FILE = "C:\Teradici\provisioning.log"
...
+    }
+}
+
 Start-Transcript -Path $LOG_FILE -Append -IncludeInvocationHeader
 
-"Not using encryption"
+if ([string]::IsNullOrWhiteSpace("${customer_master_key_id}")) {
+    "Not using encryption"
+} else {
+    "Using ecnryption key ${customer_master_key_id}"
```
Comment (Edwin-Pau) [inline modules/aws/dc/new_domain_admin_user.ps1.tpl]: Typo on encryption

File: modules/aws/dc/setup.ps1.tpl [hunk 5]
```diff
@@ -10,9 +10,28 @@ $LOG_FILE = "C:\Teradici\provisioning.log"
...
+    }
+}
+
 Start-Transcript -Path $LOG_FILE -Append -IncludeInvocationHeader
 
-"Not using encryption"
+if ([string]::IsNullOrWhiteSpace("${customer_master_key_id}")) {
+    "Not using encryption"
+} else {
+    "Using ecnryption key ${customer_master_key_id}"
```
Comment (Edwin-Pau) [inline modules/aws/dc/setup.ps1.tpl]: Typo on encryption

File: modules/aws/win-gfx/win-gfx-startup.ps1.tmpl [hunk 6]
```diff
@@ -231,7 +257,12 @@ if ($currentPrincipal.IsInRole([Security.Principal.WindowsBuiltInRole]::Administ
     "Not running as Administrator"
 }
 
-"This script is not using encryption for secrets."
+if ([string]::IsNullOrWhiteSpace("${customer_master_key_id}")) {
+    "This script is not using encryption for secrets."
+} else {
+    "Using ecnryption key ${customer_master_key_id} for secrets."
```
Comment (Edwin-Pau) [inline modules/aws/win-gfx/win-gfx-startup.ps1.tmpl]: Typo on encryption

File: modules/aws/win-std/win-std-startup.ps1.tmpl [hunk 7]
```diff
@@ -230,7 +256,12 @@ if ($currentPrincipal.IsInRole([Security.Principal.WindowsBuiltInRole]::Administ
     "Not running as Administrator"
 }
 
-"This script is not using encryption for secrets."
+if ([string]::IsNullOrWhiteSpace("${customer_master_key_id}")) {
+    "This script is not using encryption for secrets."
+} else {
+    "Using ecnryption key ${customer_master_key_id} for secrets."
```
Comment (Edwin-Pau) [inline modules/aws/win-std/win-std-startup.ps1.tmpl]: Typo on encryption

**Review by Edwin-Pau — COMMENTED** [review]
Looks good to me, just the typos. Left a comment regarding decryption.

File: modules/aws/cac/cac-startup.sh.tmpl (hunk 2 above)
Comment (shermanyin) [reply modules/aws/cac/cac-startup.sh.tmpl:32]: we could, but if it works now, we don't have to change it unless there are real benefits.

File: modules/aws/dc/dc-startup.ps1.tmpl (hunk 3 above)
Comment (shermanyin) [reply modules/aws/dc/dc-startup.ps1.tmpl]: i copy pasted this everywhere so there's going to be a lot of corrections.

**Pull request merged**